systemd_user_context.sh

```diff
@@ -5,36 +5,105 @@
 # Mainly the issue is that a desktop session needs a dedicated dbus,
 # but systemd creates a single one for every logged on user.
 #
-# Using systemd-run it is possible to create a separate systemd-user instance
-# with dedicated XDG_RUNTIME_DIR and DBUS
+# Using systemd-run or a systemd user unit file it is possible to create 
+# a separate systemd-user instance with dedicated XDG_RUNTIME_DIR and DBUS
 #
 # Main ideas in this script by mwsys.mine.bz
 
 # Units to mask in the created instance
 # Uncomment this line and fill in as appropriate for your installation
-#UNITS_TO_MASK="pipewire-pulseaudio.service pipewire-pulseaudio.socket"
+#UNITS_TO_MASK="pipewire-pulse.service pipewire-pulse.socket"
 
 # We're probably going to use file descriptor 1 for output. Stash it away
 # in fd 3 and redirect our fd 1 to stderr to talk to the user
 exec 3>&1 >&2
 
+# prepare user environment
+prepare_user_environment()
+{
+    test -f $XDG_RUNTIME_DIR/systemd/user.control/systemd-session@.service && return
+    install -dm 0700 $XDG_RUNTIME_DIR/systemd/user.control
+    # Create a unit to wait for the target pid to finish
+    {
+        echo "[Unit]"
+        echo "Description=Wait for XRDP session %i to finish"
+        echo "Requires=systemd-session@%i.service"
+        echo "ConditionPathExists=%t/systemd-session-%i/wait-for.pid"
+        echo
+        echo "[Service]"
+        echo "Type=simple"
+        echo "ExecStart=/bin/sh -c 'XRDP_STARTWM_PID=\$(cat %t/systemd-session-%i/wait-for.pid); while /bin/kill -0 \$XRDP_STARTWM_PID 2>/dev/null; do sleep 5; done'"
+        echo "ExecStopPost=/usr/bin/systemctl --user stop systemd-session@%i.service"
+        echo "ExecStopPost=rm -r %t/systemd-session-%i"
+    } >$XDG_RUNTIME_DIR/systemd/user.control/wait-for-systemd-session@.service
+
+    #Create the systemd-user session unit file.
+    {
+        echo "[Unit]"
+        echo "Description=XRDP systemd User Manager for display %i"
+        echo
+        echo "[Service]"
+        echo "Type=notify"
+        echo "ExecStart=sh %t/systemd_user_session.sh systemd-session-%i"
+    } > $XDG_RUNTIME_DIR/systemd/user.control/systemd-session@.service
+    
+    #create the systemd --user wrapper to launch systemd with a clean environment
+    echo '#/bin/sh
+test -z $XDG_RUNTIME_DIR && exit 1
+test -z $1 && exit 1
+
+SESSION_RUNTIME_DIR="$XDG_RUNTIME_DIR/$1"
+install -dm 0700 "$SESSION_RUNTIME_DIR"
+oIFS="$IFS"
+IFS="
+"
+
+# keep only absolutly needed environment variables
+for ev in `env`; do
+    evn=${ev%%=*}
+    [ "$evn" != "HOME" -a \
+      "$evn" != "SHELL" -a \
+      "$evn" != "LANG" -a \
+      "$evn" != "PATH" -a \
+      "$evn" != "SYSTEMD_EXEC_PID" -a \
+      "$evn" != "INVOCATION_ID" -a \
+      "$evn" != "NOTIFY_SOCKET" -a \
+      "$evn" != "MANAGERPID" ] \
+    && unset $evn;
```

Comment:
Quotes around "$evn"?

Reply:
I think environment variable names with spaces are not possible, so omitting then should be ok.
setting an environment without a name is impossible too afaik